panic: runtime error: invalid memory address or nil pointer dereference / storage.newReplicaCorruptionError

Please answer these questions **for each node in your cluster**:

1. What version of CockroachDB are you using (`cockroach version`)?

single node:

```
Build Vers:  go1.6
Build Tag:   beta-20160330
Build Time:  2016/04/01 02:36:55
Build Deps:
  github.com/VividCortex/ewma                c34099b489e4ac33ca8d8c5f9d29d6eeaf69f2ed
  github.com/biogo/store                     3b4c041f52c224ee4a44f5c8b150d003a40643a0
  github.com/chzyer/readline                 30b462e50b8f8a1172c3b2f780894c0db2ae2dcb
  github.com/cockroachdb/c-protobuf          4feb192131ea08dfbd7253a00868ad69cbb61b81
  github.com/cockroachdb/c-rocksdb           b80d2efe8e544bbcc4b50dec8e89f9305f5da745
  github.com/cockroachdb/c-snappy            5c6d0932e0adaffce4bfca7bdf2ac37f79952ccf
  github.com/cockroachdb/cmux                112f0506e7743d64a6eb8fedbcff13d9979bbf92
  github.com/cockroachdb/cockroach           ebf4980be08fd0fe70cb3863a6053543f816cedc
  github.com/cockroachdb/pq                  94a0880a5cc09284fe5677c86e44d5ba342912dc
  github.com/codahale/hdrhistogram           360314142131c2043d1346f197f86435b287c6da
  github.com/coreos/etcd                     83ada7232a13491f7ef94cb398d9af8b3b2cb5fb
  github.com/cpuguy83/go-md2man              2724a9c9051aa62e9cca11304e7dd518e9e41599
  github.com/dustin/go-humanize              8929fe90cee4b2cb9deb468b51fb34eba64d1bf0
  github.com/elastic/gosigar                 a19182d1f557de49dc3cef0eefa688ae6a3316d9
  github.com/elazarl/go-bindata-assetfs      57eb5e1fc594ad4b0b1dbea7b286d299e0cb43c2
  github.com/gengo/grpc-gateway              965b62d83c33af1be34fd5d70ccf814c4029a319
  github.com/gogo/protobuf                   128ef922449da3dbb6c5d942856877a1a7235fef
  github.com/golang/protobuf                 8d92cf5fc15a4382f8964b08e1f42a75c0591aa3
  github.com/google/btree                    cc6329d4279e3f025a53a83c397d2339b5705c45
  github.com/julienschmidt/httprouter        77366a47451a56bb3ba682481eed85b64fea14e8
  github.com/kr/pretty                       add1dbc86daf0f983cd4a48ceb39deb95c729b67
  github.com/kr/text                         bb797dc4fb8320488f47bf11de07a733d7233e1f
  github.com/lib/pq                          3cd0097429be7d611bb644ef85b42bfb102ceea4
  github.com/mattn/go-isatty                 56b76bdf51f7708750eac80fa38b952bb9f32639
  github.com/mattn/go-runewidth              d6bea18f789704b5f83375793155289da36a3c7f
  github.com/olekukonko/tablewriter          cca8bbc0798408af109aaaa239cbd2634846b340
  github.com/opentracing/basictracer-go      e8d818a9990870fbd77b7e03697110ee074b71f2
  github.com/opentracing/opentracing-go      01498abd158dfdbe8e251856bc7d14cb0f046fa3
  github.com/rcrowley/go-metrics             eeba7bd0dd01ace6e690fa833b3f22aaec29af43
  github.com/russross/blackfriday            b43df972fb5fdf3af8d2e90f38a69d374fe26dd0
  github.com/satori/go.uuid                  f9ab0dce87d815821e221626b772e3475a0d2749
  github.com/shurcooL/sanitized_anchor_name  10ef21a441db47d8b13ebcc5fd2310f636973c77
  github.com/spf13/cobra                     c678ff029ee250b65714e518f4f5c5cb934955de
  github.com/spf13/pflag                     7f60f83a2c81bc3c3c0d5297f61ddfa68da9d3b7
  golang.org/x/crypto                        c7e3b0ebdd409a0d024e3d71801427ab0e05fb2e
  golang.org/x/net                           1dd0d85fc0634d2dbe7f7d06d176c2ff548e6332
  golang.org/x/text                          1b466db55e0ba5d56ef5315c728216b42f796491
  google.golang.org/grpc                     9e3a674ceba65708273cf1cd8e71a1bdce68107b
  gopkg.in/inf.v0                            3887ee99ecf07df5b447e9b00d9c0b2adaa9f3e4
  gopkg.in/yaml.v1                           9f9df34309c04878acc86042b16630b0f696e1de
  /usr/local                                 1021b76c294cc3d0669b27bb22e96061fa5e50a2
```

2. What operating system and processor architecture are you using?

os x, lastest, intel i7

3. What flags/environment variables did you pass to `cockroach start`?

```
cockroach start --store=/usr/local/var/cockroach  
```

Please describe the issue you observed:


```
bash-3.2$ cockroach start --store=/usr/local/var/cockroach  
build:     beta-20160330 @ 2016/04/01 02:36:55 (go1.6)  
admin:     http://localhost:8080  
sql:       postgresql://root@localhost:26257?sslmode=disable  
logs:      /usr/local/var/cockroach/logs  
store[0]:  path=/usr/local/var/cockroach  
  
E160331 21:43:58.018688 server/server.go:301  read tcp 127.0.0.1:26257->127.0.0.1:49756: read: connection reset by peer  
panic: runtime error: invalid memory address or nil pointer dereference  
[signal 0xb code=0x1 addr=0x20 pc=0x429184d]  
  
goroutine 99 [running]:  
panic(0x5112e20, 0xc82000e100)  
        /usr/local/Cellar/go/1.6/libexec/src/runtime/panic.go:464 +0x3e6  
github.com/cockroachdb/cockroach/storage.newReplicaCorruptionError(0xc822c2f1a8, 0x3, 0x3, 0x0)  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/storage/replica.go:1882 +0xad  
github.com/cockroachdb/cockroach/storage.(*Replica).applyRaftCommand(0xc8200cfa40, 0xc820747320, 0x8, 0x9abd9d0, 0xc821d79d70, 0x9f8f3, 0x100000001, 0x1, 0x144122a91156b281, 0x0, ...)  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/storage/replica.go:1487 +0x9cb  
github.com/cockroachdb/cockroach/storage.(*Replica).processRaftCommand(0xc8200cfa40, 0xc820747320, 0x8, 0x9f8f3, 0x21, 0x100000001, 0x1, 0x144122a91156b281, 0x0, 0x100000001, ...)
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/storage/replica.go:1435 +0x2be  
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReady(0xc8200cfa40, 0x0, 0x0)  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/storage/replica.go:1299 +0xa49  
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/storage/store.go:1869 +0x35a  
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202d8af0, 0xc8201cd970)  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52  
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62  
bash-3.2$ 
```

1. What did you do?

trying to insert > 140m rows in to db with 100 connections, one at a time. It crashes fast < 1m.

2. What did you expect to see?

no crash

3. What did you see instead?

crash

when restarting;

```
$ cockroach start --store=/usr/local/var/cockroach --logtostderr  
I160331 21:53:14.918528 cli/start.go:162  [build] beta-20160330 @ 2016/04/01 02:36:55 (go1.6)  
I160331 21:53:14.918633 server/context.go:275  1 storage engine initialized  
I160331 21:53:14.918647 cli/start.go:176  starting cockroach node  
W160331 21:53:14.919417 server/server.go:104  running in insecure mode, this is strongly discouraged. See --insecure.  
W160331 21:53:14.919484 gossip/gossip.go:889  not connected to cluster; use --join to specify a connected node  
I160331 21:53:14.923849 storage/engine/rocksdb.go:120  opening rocksdb instance at "/usr/local/var/cockroach"  
W160331 21:53:14.923928 gossip/gossip.go:889  not connected to cluster; use --join to specify a connected node  
I160331 21:53:14.938505 server/node.go:372  initialized store store=1:1 ([]=/usr/local/var/cockroach): {Capacity:499055067136 Available:200868646912 RangeCount:0}  
I160331 21:53:14.938537 server/node.go:284  node ID 1 initialized  
I160331 21:53:14.938773 storage/stores.go:282  read 0 node addresses from persistent storage  
I160331 21:53:14.938925 server/node.go:493  connecting to gossip network to verify cluster ID...  
I160331 21:53:14.938947 server/node.go:514  node connected via gossip and verified as part of cluster "ac982d0e-4377-4e8e-8b82-7153010dc09e"  
I160331 21:53:14.938969 server/node.go:337  [node=1] Started node with [[]=/usr/local/var/cockroach] engine(s) and attributes []  
I160331 21:53:14.939017 server/server.go:359  starting http server at localhost:8080  
I160331 21:53:14.939032 server/server.go:360  starting grpc/postgres server at localhost:26257  
build:     beta-20160330 @ 2016/04/01 02:36:55 (go1.6)  
admin:     http://localhost:8080  
sql:       postgresql://root@localhost:26257?sslmode=disable  
logs:      /usr/local/var/cockroach/logs  
store[0]:  path=/usr/local/var/cockroach  
panic: runtime error: invalid memory address or nil pointer dereference  
[signal 0xb code=0x1 addr=0x20 pc=0x429184d]  
  
goroutine 82 [running]:  
panic(0x5112e20, 0xc82000e100)  
        /usr/local/Cellar/go/1.6/libexec/src/runtime/panic.go:464 +0x3e6  
github.com/cockroachdb/cockroach/storage.newReplicaCorruptionError(0xc8213ef1a8, 0x3, 0x3, 0x0)  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/storage/replica.go:1882 +0xad  
github.com/cockroachdb/cockroach/storage.(*Replica).applyRaftCommand(0xc820153dc0, 0xc820769e28, 0x8, 0x9aaf2c8, 0xc820ef1320, 0x2388e6, 0x100000001, 0x1, 0x1441232562bd9f68, 0x0, ...)     
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/storage/replica.go:1487 +0x9cb  
github.com/cockroachdb/cockroach/storage.(*Replica).processRaftCommand(0xc820153dc0, 0xc820769e28, 0x8, 0x2388e6, 0x1c, 0x100000001, 0x1, 0x1441232562bd9f68, 0x0, 0x100000001, ...)  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/storage/replica.go:1435 +0x2be  
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReady(0xc820153dc0, 0x0, 0x0)  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/storage/replica.go:1299 +0xa49  
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/storage/store.go:1869 +0x35a  
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820305340, 0xc8204c6260)  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52  
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker  
        /private/tmp/cockroach20160331-87796-nfu6os/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62  
bash-3.2$
```

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -1893,6 +1893,9 @@ func newReplicaCorruptionError(errs ...error) *roachpb.ReplicaCorruptionError {
 	var errMsg string
 	for i := range errs {
 		err := errs[len(errs)-i-1]
+		if err == nil {
+			continue
+		}
 		if len(errMsg) == 0 {
 			errMsg = err.Error()
 		} else {

--- a/storage/replica_test.go
+++ b/storage/replica_test.go
@@ -18,6 +18,7 @@ package storage
 
 import (
 	"bytes"
+	"errors"
 	"fmt"
 	"math"
 	"reflect"
@@ -4658,5 +4659,17 @@ func TestComputeVerifyChecksum(t *testing.T) {
 	}
 	if panicked {
 		t.Fatal("VerifyChecksum panicked")
+	}
+}
+
+func TestNewReplicaCorruptionError(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	err := newReplicaCorruptionError(errors.New("foo"), nil, errors.New("bar"), nil)
+	if cur, exp := err.String(), `error_msg:"foo (caused by bar)"`; !strings.Contains(cur, exp) {
+		t.Fatalf("expected '%s' contained in '%s'", exp, cur)
+	}
+	err = newReplicaCorruptionError(nil, nil, nil)
+	if exp, act := `error_msg:""`, err.String(); !strings.Contains(act, exp) {
+		t.Fatalf("expected '%s' contained in '%s'", exp, act)
 	}
 }

